Build/upgrade mui v4 to v5

timheilman opened the pull request:
Closes #1418 and partial of #1508

I depended heavily on the codemods, especially jss-to-styled, because I am not a CSS expert.

timheilman committed:
build: yarn add @mui/material @mui/styles ; yarn add @mui/lab ; yarn add @mui/icons-material

timheilman committed:
build: yarn add @emotion/react @emotion/styled

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/preset-safe src

timheilman committed:
build: empty commit, for: npx @mui/codemod@latest v5.0.0/variant-prop src

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/link-underline-hover src

timheilman committed:
build: npx prettier --write src backend cypress scripts

timheilman committed:
build: yarn remove @material-ui/core @material-ui/icons @material-ui/lab

timheilman committed:
build: rm patches/@material-ui+core+4.12.4.patch

timheilman committed:
build: update renovate.json to use @mui not @material-ui

timheilman committed:
build: update ThemeProvider import, per

https://mui.com/material-ui/migration/v5-style-changes/\#update-themeprovider-import

timheilman committed:
build: update EventType for Tabs onChange event, per

https://mui.com/material-ui/migration/v5-component-changes/\#update-event-type-typescript-5

timheilman committed:
fix: Two tabs problems:

First, the color changes indicated at https://mui.com/material-ui/migration/v5-component-changes/#update-default-indicatorcolor-and-textcolor-prop-values

Second, remove the the trailing "px" a number rather than string when passing values to the react-virtualized List component.  See
https://mui.com/material-ui/migration/v5-style-changes/\#%E2%9C%85-remove-px-suffix

timheilman committed:
fix: transaction amount was falling off right edge

timheilman committed:
fix: indirection of onChange prop of Slider via ownerState

this was causing the slider-based cypress tests to fail.

timheilman committed:
fix: remove zIndex from AppBar to fix drawer issue

appBar was still over the date picker in mobile screen size, failing one of the cypress:run:mobile tests.  See also
https://mui.com/material-ui/migration/v5-component-changes/\#fix-z-index-issues

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/containers

followed by lint fixes and manual review

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/BankAccountForm.tsx

linting and review

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/CommentForm.tsx

review, lint

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/MainLayout.tsx

review, remove class for root and use ampersand instead, adding a flexGrow:1 to get styling as before

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/NavBar.tsx,

review, appBarShift was incorrectly considered a child of root; fixed that with removal of space between & and ., lint

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/NavDrawer.tsx

review, lint, no issues

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/NotificationListItem.tsx

review, lint, no issues

timheilman committed:
build:  npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/SignInForm.tsx

lint, cast StyledContainer to typeof Container per https://github.com/mui/material-ui/issues/15695\#ref-commit-7fd1a29

timheilman committed:
build:  npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/SignUpForm.tsx

lint, cast StyledContainer to typeof Container per https://github.com/mui/material-ui/issues/15695\#ref-commit-7fd1a29

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/SkeletonList.tsx

lint, review, no issues

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/TransactionAmount.tsx

lint, remove space between & and . because classes are conditionally applied

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/TransactionCreateStepOne.tsx

lint and review, no issues

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/TransactionCreateStepTwo.tsx

lint, review, no issues

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/TransactionCreateStepThree.tsx

lint, review, no issues

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/TransactionDateRangeFilter.tsx

lint, review, no issues

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/TransactionDetail.tsx

lint, review, no issues

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/TransactionItem.tsx

lint, review, problems with SmallAvatar, fixed manually

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/TransactionList.tsx

lint, review, no issues

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/TransactionListAmountRangeFilter.tsx

lint, review, did not work really at all, fixed by using sx prop rather than classes at all

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/TransactionListFilters.tsx

lint, review, no issues

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/TransactionTitle.tsx

lint, review, no issues

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components/UserSettingsForm.tsx

lint, review, form did not want to play nice with classes, fixed with sx for submit button and styled wrapping div for form

timheilman committed:
build: npx @mui/codemod@latest v5.0.0/jss-to-styled src/components

had apparently accidentally failed-to-checkin TransactionInfiniteList and userSearchForm, but both were reviewed and had no issues

timheilman committed:
build: yarn remove @mui/styles and some

minor style fixups to be rid of the dependency causing the peer dependency warning

MikeMcC399 reviewed (changes requested):
@timheilman 

It's great you've picked up this migration. Well done! 👍🏻 

The `data/database.json` should not be part of the PR.  There is something unclean in the repo causing the file to get changed, when it ought to be static. I have noticed this before after running tests but I didn't investigate it further. 🙁

timheilman committed:
fix: oops did not mean to modify database.json

timheilman commented:
mui v5, o how it needs a hero...

MikeMcC399 reviewed (changes requested):
You've requested a review. The PR is however showing conflicts, so these would need to be resolved.

timheilman committed:
Merge branch 'develop' into build/upgrade-mui-v4-to-v5

timheilman committed:
build: remove @ts-expect-error annotations no longer used after yarn.lock

merge resolution

timheilman committed:
test: issue 1278 re-enabling test that had failed on firefox,

which is not failing after the upgrade to MUI v5.  See
https://github.com/cypress-io/cypress-realworld-app/issues/1278

timheilman committed:
style: npm run prettier

timheilman commented:
@MikeMcC399 OK merge conflicts are resolved.

Also, in the meantime I discovered https://github.com/cypress-io/cypress-realworld-app/issues/1278 . Indeed this test is passing in firefox after this upgrade to MUI v5.

This upgrade was a lot of work.  A lot of slow, "grunt" work -- updating a page, verifying it still rendered correctly manually, verifying it still passed tests.  The focus was: eliminate the build warnings in #1418 .  _The build warning is eliminated_.  Whatever further changes may need to be made, this PR is a step in the right direction, since MUI v4 is dying.

MikeMcC399 reviewed (approved):
@timheilman 

Congratulations on the successful migration! A look at the number of source files you had to change goes to show how much work this must have been. 💪🏻

AtofStryker commented:
really nice work on this @timheilman ! I reviewed locally and things look as expected on my end. I opened https://github.com/cypress-io/cypress-realworld-app/pull/1550 so we can run percy to see visual regressions. I will close it of course in favor of your PR. The only purpose it serves is to run a few additional jobs we don't normally run for external contributors.

jennifer-shehane commented:
@timheilman The percy snapshots in the forked PR seem to be highlighting some changes in styles on the Transaction View page which don't look desirable. Could you take a look?

![Screenshot 2024-05-20 at 3 14 24 PM](https://github.com/cypress-io/cypress-realworld-app/assets/1271364/74a5bcc4-a479-448e-ab4c-8cc44206be4f)
![Screenshot 2024-05-20 at 3 14 08 PM](https://github.com/cypress-io/cypress-realworld-app/assets/1271364/1563d440-350b-4576-ba60-353e35c416fc)

timheilman commented:
I'm unfamiliar with percy but I will try to understand what's going on and see what I can do when I get some free cycles.

jennifer-shehane commented:
@timheilman Percy is just a visual regression testing framework. So it will take screenshots before a commit and screenshots after a commit and compare them for the entire app where you tell it to take screenshots. Unfutunately contributors don't have access to our Percy account so can't directly review the screenshots. 

The screenshots after this change indicate that the 'Transaction View page' looks different after this change in this PR, which I don't think is intentional. 

You can look at the change within Test Replay to narrow down the cause. You can see the avatars are in a different place in those.
- The [Test Replay before this change](https://cloud.cypress.io/projects/7s5okt/runs/6399/test-results/ee89c42b-73ca-4428-a3d0-af0301b8b220/replay?actions=%5B%5D&browsers=%5B%5D&groups=%5B%5D&isFlaky=%5B%5D&modificationDateRange=%7B%22startDate%22%3A%221970-01-01%22%2C%22endDate%22%3A%222038-01-19%22%7D&orderBy=EXECUTION_ORDER&oses=%5B%5D&specs=%5B%7B%22label%22%3A%22cypress%2Ftests%2Fui%2Ftransaction-view.spec.ts%22%2C%22value%22%3A%22%5B%5C%223a57ac4f-df3f-42a9-b39f-9618f07d7618%5C%22%5D%22%7D%5D&statuses=%5B%5D&testingTypesEnum=%5B%5D&ts=1716217148999.6&att=1)
- The [Test Replay of this change](https://cloud.cypress.io/projects/7s5okt/runs/6398/test-results/bcd8cc37-79e4-4fb9-ac41-1e52ef70d26a/replay?actions=%5B%5D&browsers=%5B%5D&groups=%5B%5D&isFlaky=%5B%5D&modificationDateRange=%7B%22startDate%22%3A%221970-01-01%22%2C%22endDate%22%3A%222038-01-19%22%7D&orderBy=EXECUTION_ORDER&oses=%5B%5D&specs=%5B%7B%22label%22%3A%22cypress%2Ftests%2Fui%2Ftransaction-view.spec.ts%22%2C%22value%22%3A%22%5B%5C%222986aeed-5306-4cf5-940f-bfc47ab62e49%5C%22%5D%22%7D%5D&statuses=%5B%5D&testingTypesEnum=%5B%5D&ts=1715975087805.4&att=1)

timheilman committed:
Merge remote-tracking branch 'origin/develop' into build/upgrade-mui-v4-to-v5

timheilman commented:
OK, I've made some small progress on the TransactionDetail unintended layout change.
The offending commit is 552005ba757d366545fb772e9e8c8f524ae99445, resulting from the command:

`npx @mui/codemod@latest v5.0.0/preset-safe src`

timheilman commented:
Looks like [the removal of alignItems and justifyContent props from the Grid component](https://mui.com/material-ui/migration/v5-component-changes/#%E2%9C%85-remove-align-and-justify-props-and-classes) is the problem.

timheilman commented:
> Looks like [the removal of alignItems and justifyContent props from the Grid component](https://mui.com/material-ui/migration/v5-component-changes/#%E2%9C%85-remove-align-and-justify-props-and-classes) is the problem.

Actually, not.  My reading of that is that the props are still OK; just that MUI system took them over from the component.  And CSS inspection shows them applied.

For the avatar group root class, two CSS properties differ: a `flex-direction: "row-reverse"` is coming in, and `font-size:` is `16px` rather than `14px`.

timheilman committed:
fix: unset flex-direction which is newly set to row-reverse for

AvatarGroup in mui5.

timheilman committed:
fix: undo parent commit fix because avatars were backward and

no longer overlapping.  Instead eat up space to right of avatars.

timheilman commented:
the flex-direction issue looks like it was from a change in the `<AvatarGroups>` component, and a fix is pushed.  The font-size issue, and other css issues with text input elements seem to be related to changes in defaults from the jss styling engine to the emotion styling engine.  Next I'll look for how to set emotion defaults to the previous jss defaults.

timheilman committed:
fix: get body font size back down to 14, establish knobs for

modifying input styles

AtofStryker commented:
I reran the percy jobs and the diff looks pretty close to me. I would say this is acceptable. @timheilman did the default theme colors change slightly?
![](https://github.com/cypress-io/cypress-realworld-app/assets/3980464/73d4db4d-6207-4d90-badc-e00b46a1d468)

It also looks like there is a failure in the mobile viewport tests? Though I don't see the error on my updated mirror PR https://github.com/cypress-io/cypress-realworld-app/pull/1550 so I am wondering if it is legit

timheilman commented:
I'm back into my work week and have to table this in favor of paid work for now.  I'm actually cutting my teeth here -- I've never used MUI v4 _nor_ v5 before, nor any theming engine for that matter.  It does look like the palette changed, again probably because of the jss->emotion switch, but I would need to debug with the css developer tools to find out why.  Of course the test failures need looking into.

If anyone else wants to take over the branch/PR, that's fine by me.  Or I will look at it again when I get the time.

timheilman commented:
See https://github.com/replayio-public/cypress-realworld-app/pull/113

timheilman commented:
oops that was a fork!

timheilman committed:
Merge remote-tracking branch 'origin/develop' into build/upgrade-mui-v4-to-v5

timheilman commented:
> It also looks like there is a failure in the mobile viewport tests?

I am able to reproduce these failures locally so I think they're legit.

timheilman commented:
OK it looks to me like, under the mobile viewport, the test `ui/transaction-view.spec.ts` is failing due to [issue 29776](https://github.com/cypress-io/cypress/issues/29776).

timheilman committed:
test: apply force:true to clicks failing due to issue #29776

timheilman commented:
I'll check back on this and if the font sizes and theming colors remain blockers to merging investigate how to fix them then. I'm not happy about the `{ force: true }` fix to the failing mobile viewport tests, but as far as I can tell those test failures are exhibiting a genuine bug in cypress.

AtofStryker committed:
Merge branch 'develop' of github.com:cypress-io/cypress-realworld-app into mui5

AtofStryker commented:
@jennifer-shehane the changes in percy look acceptable to me 
![](https://github.com/user-attachments/assets/801b52f5-b188-43f7-b541-6184c3572b03)

AtofStryker commented on cypress/tests/ui/transaction-view.spec.ts:
```diff
@@ -58,7 +60,8 @@ describe("Transaction View", function () {
   it("comments on a transaction", function () {
-    cy.getBySelLike("transaction-item").first().click();
+    // { force: true } is a workaround for https://github.com/cypress-io/cypress/issues/29776
+    cy.getBySelLike("transaction-item").first().click({ force: true });
```
this seems ok until we can address the issue

AtofStryker commented:
@timheilman thank you so much for doing this and sorry for the delay in getting this in!
